proposals/nnnn-vector.md

```diff
@@ -0,0 +1,797 @@
+  withUnsafePointer(to: section.sectname) {
+    // This is unsafe! 'sectname' isn't guaranteed to have a null byte
+    // indicating the end of the C string!
+    String(cString: $0)
+  }
+}
+
+func iterateSectionNameBytes(_ section: section_64) {
+  withUnsafeBytes(to: section.sectname) {
+    for byte in $0 {
+      ...
+    }
+  }
+}
+```
+
+Having to resort to using very unsafe API to do anything useful with imported C
+arrays is not something a memory safe language like Swift should be in the
+business of. `Vector` allows us to clean up this sore spot in a very elegant and
+much more importantly, safe way.
+
+To preserve source compatibility, we will be introducing a new imported field
+for C structs that contain C arrays. This field will be named according to the
+imported C array field's name by appending `Vector` to the end of it:
+`<C name>Vector`.
+
+```c
+struct section_64 {
+  char sectname[16];
+  char segname[16];
+  uint64_t addr;
+  uint64_t size;
+  uint32_t offset;
+  uint32_t align;
+  ...
+};
+```
+
+Would be imported as:
+
+```swift
+struct section_64 {
+  let sectname: (CChar, CChar, CChar, CChar, CChar, CChar, ... 10 more times)
+  let sectnameVector: Vector<16, CChar>
+  let segname: (CChar, CChar, CChar, CChar, CChar, CChar, ... 10 more times)
+  let segnameVector: Vector<16, CChar>
+  let addr: UInt64
+  let size: UInt64
+  let offset: UInt32
+  let align: UInt32
+  ...
+}
+```
+
+If the C struct already contained a field named `sectnameVector`, then we
+wouldn't generate a `Vector` field for `sectname`. However, one should be able
+to mitigate this by annotating the C field directly with `swift_vector_name`:
+
+```c
+#define SWIFT_VECTOR_NAME(X) __attribute__((__swift_vector_name__(#X)))
+
+struct section_64 {
+  SWIFT_VECTOR_NAME(sectname_vector)
+  char sectname[16];
+
+  SWIFT_VECTOR_NAME(segname_vector)
+  char segname[16];
+
+  uint64_t addr;
+  uint64_t size;
+  uint32_t offset;
+  uint32_t align;
+  ...
+};
+```
+
+Note, that while the imported C struct looks like it has more members, its
+memory layout will be unaffected.
+
+In a future Swift language mode, we plan to disable this behavior and drop
+importing C array fields as tuples and only as `Vector` with the same name from
+the C struct. This approach lets folks who are ready to migrate their codebase
+using a C interface to use `Vector` entirely, while still allowing dependencies
+or dependents to use the old behavior with the same C interface.
+
+## Source compatibility
+
+`Vector` is a brand new type in the standard library, so source should still be
+compatible.
+
+Given the name of this type however, we forsee this clashing with existing user
+defined types named `Vector`. This isn't a particular issue though because the
+standard library has special shadowing rules which prefer user defined types by
+default. Which means in user code with a custom `Vector` type, that type will
+always be preferred over the standard library's `Swift.Vector`.
```

Comment:
I'm assuming it does, but can you clarify if "always" means that `MyLib.Vector<T>` will be preferred over `Swift.Vector<N, T>` even if I write `Vector<3, Int>`?

Reply:
This is a good find and it seems the behavior is yes, it will \_always\_ be preferred. I'll make a note of it in the proposal.